### PR TITLE
DOC: replace dream with sexp in first program

### PR DESCRIPTION
**Pull request by heathhenley**
Related to #1962 

What does everyone think about something like this? I tested on windows and it works, the example is just pulled from Yojson's readme, but we could make it 'learning ocaml' specific if you prefer. 

Two things:
- after installing the package using opam, I had to run the opam env update described here: https://ocaml.org/docs/installing-ocaml#2-create-an-opam-switch (not creating the switch just update the env). I'm wondering if that should be mentioned in the tutorial? Opam does tell you to run `eval $(opam env)` but that only helps if you're running linux based...
- is there a better way to describe the type that Yojson returns that is more instructive for beginners?

### CODE REVIEW

**Comment by F-Loyer** [comment]
I guess I was too fast when suggesting `yojson`. It is small, not system-dependant, useful... but it uses polymorphic variants which I don't think they are introduced in the learning/language section.

sexplib belongs to the same kind of library, but Sexp instead of JSON. The result type doesn't use polymorphic variants, but simple ADT.

**Comment by heathhenley** [comment]
Ah, no worries I'll give that one a shot instead. I went with JSON because it was more familiar, now I get to learn what S expression is 🤣

**Comment by F-Loyer** [comment]
"is there a better way to describe the type that Yojson returns that is more instructive "

Some programs need a configuration file. Then if we read it, it can be convenient to parse it, especially if the config is complex (variable assigned to list...). Then a program which extract a value could demonstrate an interest.

JSON and Sexp both permit this kind of configuration.

Note Polymorphic variants ADT are very close to the usual ADT... Perhaps we could use them.

(You already have typed a Sexp... `dune` file are like this).

File: data/tutorials/getting-started/1_02_your_first_ocaml_program.md [hunk 1]
```diff
@@ -241,58 +241,43 @@ A more detailed introduction to modules can be found at [Modules](/docs/modules)
 
 OCaml has an active community of open-source contributors. Most projects are available using the opam package manager, which you installed in the [Install OCaml](/docs/up-and-ready) tutorial. The following section shows you how to install and use a package from opam's open-source repository.
 
-To illustrate this, let's turn our `hello` project into a web server using [Anton Bachin](https://github.com/aantron)'s [Dream](https://aantron.github.io/dream/) web framework. First, update the package list for Opam, by running `opam update`. Then, install the `dream` package with this command:
+To illustrate this, let's update our `hello` project to parse a string into
+JSON using [Yojson](https://ocaml-community.github.io/yojson/yojson/index.html). First, update the package list for Opam, by running `opam update`. Then, install the `Yojson` package with this command:
```
Comment (christinerose) [inline data/tutorials/getting-started/1_02_your_first_ocaml_program.md]:
```suggestion
JSON using [Yojson](https://ocaml-community.github.io/yojson/yojson/index.html). First, update the package list for opam, by running `opam update`. Then, install the `Yojson` package with this command:
```

**Comment by heathhenley** [comment]
Here's another option based on the suggestion of Sexplib - let me know what you all think

File: data/tutorials/getting-started/1_02_your_first_ocaml_program.md [hunk 2]
```diff
@@ -239,45 +239,41 @@ A more detailed introduction to modules can be found at [Modules](/docs/modules)
 
 ## Installing and Using Modules From a Package
 
-OCaml has an active community of open-source contributors. Most projects are available using the opam package manager, which you installed in the [Install OCaml](/docs/up-and-ready) tutorial. The following section shows you how to install and use a package from opam's open-source repository.
+OCaml has an active community of open-source contributors. Most projects are available using the Opam package manager, which you installed in the [Install OCaml](/docs/up-and-ready) tutorial. The following section shows you how to install and use a package from Opam's open-source repository.
```
Comment (christinerose) [inline data/tutorials/getting-started/1_02_your_first_ocaml_program.md]:
```suggestion
OCaml has an active community of open-source contributors. Most projects are available using the opam package manager, which you installed in the [Install OCaml](/docs/up-and-ready) tutorial. The following section shows you how to install and use a package from opam's open-source repository.
```
The official format for opam is all lowercase unless it's at the beginning of a sentence.

File: data/tutorials/getting-started/1_02_your_first_ocaml_program.md [hunk 3]
```diff
@@ -239,45 +239,41 @@ A more detailed introduction to modules can be found at [Modules](/docs/modules)
...
 
-To illustrate this, let's update our `hello` project to parse a string into
-JSON using [Yojson](https://ocaml-community.github.io/yojson/yojson/index.html). First, update the package list for opam, by running `opam update`. Then, install the `Yojson` package with this command:
+To illustrate this, let's update our `hello` project to parse a string containing an S-expression and print back to a string using [Sexplib](https://github.com/janestreet/sexplib). First, update the package list for `Opam`, by running `opam update`. Then, install the `Sexplib` package with this command:
```
Comment (christinerose) [inline data/tutorials/getting-started/1_02_your_first_ocaml_program.md]:
```suggestion
To illustrate this, let's update our `hello` project to parse a string containing an S-expression and print back to a string using [Sexplib](https://github.com/janestreet/sexplib). First, update the package list for `opam`, by running `opam update`. Then, install the `Sexplib` package with this command:
```

File: data/tutorials/getting-started/1_02_your_first_ocaml_program.md [hunk 4]
```diff
@@ -239,45 +239,41 @@ A more detailed introduction to modules can be found at [Modules](/docs/modules)
...
 ```shell
-$ opam install yojson
+$ opam install sexplib
 ```
 
-Next, define a string containing some valid JSON in `bin/main.ml` and parse it
-using the `Yojson.Safe.from_string` function:
+Next, define a string containing a valid S-expression in `bin/main.ml`. Parse
+it into a S-expression with  the  `Sexplib.Sexp.of_string` function, and then
```
Comment (christinerose) [inline data/tutorials/getting-started/1_02_your_first_ocaml_program.md:250]:
```suggestion
it into an S-expression with  the `Sexplib.Sexp.of_string` function, and then
```

File: data/tutorials/getting-started/1_02_your_first_ocaml_program.md [hunk 5]
```diff
@@ -239,45 +239,41 @@ A more detailed introduction to modules can be found at [Modules](/docs/modules)
...
 
-let json = Yojson.Safe.from_string json_string
+(* Do something with the Sexp ... *)
 
-let () = Format.printf "Parsed to %a" Yojson.Safe.pp json
+(* Convert back to a string to print *)
+let () = Printf.printf "%s\n" (Sexplib.Sexp.to_string exp1)
 ```
-The JSON string is parsed into a [JSON type](https://ocaml-community.github.io/yojson/yojson/Yojson/Safe/index.html) so that you can use it in your program. Refer to the [Yojson documentation](https://ocaml-community.github.io/yojson/yojson/index.html) for more information.
+The string you entered representing a valid S-expression is parsed into `List`of `List`s of `strings` so that you can use it in your program. Refer to the [Sexplib documentation](https://github.com/janestreet/sexplib) for more information.
```
Comment (christinerose) [inline data/tutorials/getting-started/1_02_your_first_ocaml_program.md]:
```suggestion
The string you entered representing a valid S-expression is parsed into `List` of `List`s of `strings` so that you can use it in your program. Refer to the [Sexplib documentation](https://github.com/janestreet/sexplib) for more information.
```

**Review by christinerose — COMMENTED** [review]
A few suggestions and typos.

File: data/tutorials/getting-started/1_02_your_first_ocaml_program.md (hunk 5 above)
Comment (heathhenley) [reply data/tutorials/getting-started/1_02_your_first_ocaml_program.md]: Thanks for the corrections! A pushed a change to the type definition thanks to the clarification by @F-Loyer [ here ](https://github.com/ocaml/ocaml.org/commit/483d66ce53b32481be10cf3e989fcbf1d0ed2b1e) so this suggestion looks outdated. Though based on their  correction (that I just noticed) this probably needs another tweak to be more correct...

Comment (F-Loyer) [reply data/tutorials/getting-started/1_02_your_first_ocaml_program.md]: Beware, I was wrong in my first comment (I have edited). It is not a list of atom (string) or sexp, but an atom (string) or a list of s-exp.

File: data/tutorials/getting-started/1_02_your_first_ocaml_program.md [hunk 6]
```diff
@@ -241,58 +241,41 @@ A more detailed introduction to modules can be found at [Modules](/docs/modules)
...
-- `get "/"` declares a route, HTTP GET requests to the `/` path are handled by the provided function.
-- `(fun (_ : request) -> ...)` this a handler function. The typed pattern `(_ : request)` means the data from the request is ignored.
-- `html Hello.En.v` has type `response promise`. When data inside the promise becomes available, the server will send it, too. In our case, it is immediately available as it is a static constant stored in memory.
-
-In summary, this is telling: “run a web server responding with the content of `Hello.En.v` to requests to the '/' path”
+(* Convert back to a string to print *)
+let () = Printf.printf "%s\n" (Sexplib.Sexp.to_string exp1)
+```
+The string you entered representing a valid S-expression is parsed into
+an S-expression type, which is defined as `List` of items that can be either an
```
Comment (christinerose) [inline data/tutorials/getting-started/1_02_your_first_ocaml_program.md]:
```suggestion
an S-expression type, which is defined as a `List` of items that can be either an
```
Should there be an "a" here?

**Comment by christinerose** [comment]
I see several of my suggestions are now outdated, so I'll go through it again when the content is set. 🙂

**Comment by heathhenley** [comment]
I updated the type def - I think it's ready for a re-review / input whether it's a good enough example to switch for dream.

File: data/tutorials/getting-started/1_02_your_first_ocaml_program.md (hunk 6 above)
Comment (christinerose) [reply data/tutorials/getting-started/1_02_your_first_ocaml_program.md]: Should this read "defined as a `List` of items" rather than "defined as `List` of items`?

**Review by christinerose — APPROVED** [review]
Looks great!

File: data/tutorials/getting-started/1_02_your_first_ocaml_program.md [hunk 7]
```diff
@@ -241,58 +241,41 @@ A more detailed introduction to modules can be found at [Modules](/docs/modules)
 
 OCaml has an active community of open-source contributors. Most projects are available using the opam package manager, which you installed in the [Install OCaml](/docs/up-and-ready) tutorial. The following section shows you how to install and use a package from opam's open-source repository.
 
-To illustrate this, let's turn our `hello` project into a web server using [Anton Bachin](https://github.com/aantron)'s [Dream](https://aantron.github.io/dream/) web framework. First, update the package list for Opam, by running `opam update`. Then, install the `dream` package with this command:
+To illustrate this, let's update our `hello` project to parse a string containing an S-expression and print back to a string using [Sexplib](https://github.com/janestreet/sexplib). First, update the package list for opam by running `opam update`. Then, install the `Sexplib` package with this command:
```
Comment (cuihtlauac) [inline data/tutorials/getting-started/1_02_your_first_ocaml_program.md]:
```suggestion
To illustrate this, let's update our `hello` project to parse a string containing an [S-expression](https://en.wikipedia.org/wiki/S-expression) and print back to a string, both using [Sexplib](https://github.com/janestreet/sexplib). First, update the package list for opam by running `opam update`. Then, install the `Sexplib` package with this command:
```

File: data/tutorials/getting-started/1_02_your_first_ocaml_program.md [hunk 8]
```diff
@@ -241,58 +241,41 @@ A more detailed introduction to modules can be found at [Modules](/docs/modules)
...
 ```lisp
 (executable
  (public_name hello)
  (name main)
- (libraries hello dream))
+ (libraries hello sexplib))
 ```
 
-Launch the server from a new terminal.
+Finally, execute as before:
```
Comment (cuihtlauac) [inline data/tutorials/getting-started/1_02_your_first_ocaml_program.md:277]:
```suggestion
**Fun fact**: Dune configuration files are S-expressions.

Finally, execute as before:
```

**Review by cuihtlauac — CHANGES_REQUESTED** [review]
This looks good to me, only small changes suggested.

**Comment by cuihtlauac** [comment]
@jonahbeckford: What's your view on this PR? It should be good for Windows?

**Review by cuihtlauac — APPROVED** [review]

**Comment by jonahbeckford** [comment]
It works on Windows, yes. (I can't think of a more portable example, although I'm not enthused about the example itself. Better than the alternate so LGTM)
